File: src/main/java/com/machfour/macros/queries/MealQueries.java
package com.machfour.macros.queries;

import com.machfour.macros.core.ColumnData;
import com.machfour.macros.core.ObjectSource;
import com.machfour.macros.core.Schema;
import com.machfour.macros.objects.Food;
import com.machfour.macros.objects.FoodPortion;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.storage.MacrosDataSource;
import com.machfour.macros.util.DateStamp;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.sql.SQLException;
import java.util.ArrayList;
import java.util.Collections;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import static com.machfour.macros.core.MacrosUtils.getOrDefault;

public class MealQueries {
    public static void saveFoodPortions(@NotNull MacrosDataSource ds, @NotNull Meal m) throws SQLException {
        for (FoodPortion fp : m.getFoodPortions()) {
            if (!fp.getObjectSource().equals(ObjectSource.DATABASE)) {
                Queries.saveObject(ds, fp);
            }
        }
    }

    @NotNull
    public static Meal getOrCreateMeal(@NotNull MacrosDataSource ds, @NotNull DateStamp day, @NotNull String name) throws SQLException {
        Map<Long, Meal> mealsForDay = getMealsForDay(ds, day);
        Meal nameMatch = findMealWithName(mealsForDay, name);
        if (nameMatch != null) {
            return nameMatch;
        } else {
            ColumnData<Meal> newMealData = new ColumnData<>(Meal.table());
            newMealData.put(Schema.MealTable.DAY, day);
            newMealData.put(Schema.MealTable.NAME, name);
            Meal newMeal = Meal.factory().construct(newMealData, ObjectSource.USER_NEW);
            Queries.saveObject(ds, newMeal);
            // get it back again, so that it has an ID and stuff
            mealsForDay = getMealsForDay(ds, day);
            nameMatch = findMealWithName(mealsForDay, name);
            assert (nameMatch != null) : "didn't find saved meal in meals for its day";
            return nameMatch;
        }

    }

    // finds whether there is a 'current meal', and returns it if so.
    // defined as the most recently modified meal created for the current date
    // if no meals exist for the current date, returns null
    @Nullable
    public static Meal getCurrentMeal(@NotNull MacrosDataSource ds) throws SQLException {
        Map<Long, Meal> mealsForDay = getMealsForDay(ds, DateStamp.forCurrentDate());
        if (mealsForDay.isEmpty()) {
            return null;
        } else {
            // most recently modified -> largest modification time -> swap compare order
            return Collections.max(mealsForDay.values(),
                    (Meal a, Meal b) -> Long.compare(b.getStartTime(), a.getStartTime()));

        }
    }

    @NotNull
    public static Map<Long, Meal> getMealsForDay(@NotNull MacrosDataSource ds, @NotNull DateStamp day) throws SQLException {
        List<Long> mealIds = getMealIdsForDay(ds, day);
        return getMealsById(ds, mealIds);
    }

    @NotNull
    public static List<Long> getMealIdsForDay(MacrosDataSource ds, @NotNull DateStamp day) throws SQLException {
        return Queries.selectColumn(ds, Schema.MealTable.instance(), Schema.MealTable.ID, Schema.MealTable.DAY, Collections.singletonList(day));
        // TODO: need "DATE(" + Meal.Column.DAY + ") = DATE ( ? )"; ???
    }

    /* The get<Object>By(Id|Key) functions construct objects for all necessary entities that match the query,
     * as well as all other entities referenced by them.
     * For example, getMealsForDay constructs all of the MealTable objects for one particular day,
     * along with their FoodPortions, their Foods, and all of the Servings of those Foods.
     * It's probably worth caching the results of these!
     */
    @Nullable
    public static Meal getMealById(MacrosDataSource ds, @NotNull Long id) throws SQLException {
        Map<Long, Meal> resultMeals = getMealsById(ds, Collections.singletonList(id));
        return getOrDefault(resultMeals, id, null);
    }

    @NotNull
    public static Map<Long, Meal> getMealsById(MacrosDataSource ds, @NotNull List<Long> mealIds) throws SQLException {
        if (mealIds.isEmpty()) {
            return Collections.emptyMap();
        }
        List<Long> foodIds = getFoodIdsForMeals(ds, mealIds);
        Map<Long, Meal> meals = QueryHelpers.getRawMealsById(ds, mealIds);
        // this check stops an unnecessary lookup of all foods, which happens if no IDs are passed
        // into getFoodsById;
        if (!foodIds.isEmpty()) {
            Map<Long, Food> foodMap = FoodQueries.getFoodsById(ds, foodIds);
            for (Meal meal : meals.values()) {
                QueryHelpers.applyFoodPortionsToRawMeal(ds, meal, foodMap);
            }
        }
        return meals;
    }

    @NotNull
    public static List<Long> getFoodIdsForMeals(MacrosDataSource ds, List<Long> mealIds) throws SQLException {
        return ds.selectColumn(FoodPortion.table(), Schema.FoodPortionTable.FOOD_ID, Schema.FoodPortionTable.MEAL_ID, mealIds, true);
    }

    @Nullable
    public static Meal findMealWithName(Map<Long, Meal> mealMap, @NotNull String name) {
        Meal found = null;
        for (Meal m : mealMap.values()) {
            if (name.equals(m.getName())) {
                found = m;
            }
        }
        return found;
    }
}

File: src/main/java/com/machfour/macros/queries/MealQueries.kt
package com.machfour.macros.queries

import com.machfour.macros.orm.schema.FoodPortionTable
import com.machfour.macros.orm.schema.MealTable
import com.machfour.macros.entities.Food
import com.machfour.macros.entities.FoodPortion
import com.machfour.macros.entities.Meal
import com.machfour.macros.queries.FoodQueries.getFoodsById
import com.machfour.macros.persistence.MacrosDatabase
import com.machfour.macros.util.DateStamp
import java.sql.SQLException

object MealQueries {

    @Throws(SQLException::class)
    fun getMealsForDay(ds: MacrosDatabase, day: DateStamp): Map<Long, Meal> {
        val mealIds = getMealIdsForDay(ds, day)
        return getMealsById(ds, mealIds)
    }

    // assumes unique meal name per day
    @Throws(SQLException::class)
    fun getMealForDayWithName(ds: MacrosDatabase, day: DateStamp, name: String): Meal? {
        val mealIds = getMealIdsForDay(ds, day)
        return getMealsById(ds, mealIds).filter { it.value.name == name }.values.firstOrNull()
    }

    @Throws(SQLException::class)
    fun getMealIdsForDay(ds: MacrosDatabase, day: DateStamp): List<Long> {
        val ids = CoreQueries.selectSingleColumn(ds, MealTable.instance, MealTable.ID) {
            where(MealTable.DAY, listOf(day))
        }
        return ids.map { requireNotNull(it) { "Null meal ID encountered : $it" } }
    }

    @Throws(SQLException::class)
    fun getMealById(ds: MacrosDatabase, id: Long): Meal? {
        val resultMeals = getMealsById(ds, listOf(id))
        return resultMeals.getOrDefault(id, null)
    }

    // Creates a new foodportion from the old one, with an updated meal ID.
    // Old meal will have the old foodPortion removed and provided no other references to it are kept,
    // it will be garbage collected.
    // newMeal will have the newly constructed object (with the correct ID) added.
    // if the food portion already belongs to newMeal, nothing is done
    //@Throws(SQLException::class)
    //fun moveFoodPortion(ds: MacrosDataSource, fp: FoodPortion, newMeal: Meal) {
    //    if (fp.meal != newMeal) {
    //        val editedFp = MacrosBuilder(fp).run {
    //            setField(FoodPortionTable.MEAL_ID, newMeal.id)
    //            build()
    //        }
    //        WriteQueries.saveObject(ds, editedFp)
    //        val updatedFp = QueryHelpers.getRawObjectsByIds(ds, FoodPortion.table, listOf(fp.id))

    //        assert(updatedFp.size == 1) { "more than 1 new food portion returned" }
    //        processRawFoodPortions(ds, newMeal, updatedFp, mapOf(fp.foodId to fp.food))

    //        fp.removeFromMeal()
    //    }
    //}

    @Throws(SQLException::class)
    private fun getRawFoodPortionsForMeal(ds: MacrosDatabase, meal: Meal): Map<Long, FoodPortion> {
        val fpIds = CoreQueries.selectNonNullColumn(ds, FoodPortion.table, FoodPortionTable.ID) {
            where(FoodPortionTable.MEAL_ID, listOf(meal.id))
            distinct()
        }
        return if (fpIds.isNotEmpty()) {
            RawEntityQueries.getRawObjectsByIds(ds, FoodPortion.table, fpIds)
        } else {
            emptyMap()
        }
    }

    @Throws(SQLException::class)
    private fun processRawFoodPortions(meal: Meal, fpMap: Map<Long, FoodPortion>, foodMap: Map<Long, Food>) {
        // sort by create time - but if adding to existing meals, they are added to the end
        val fpList = fpMap.values.toList().sortedBy { it.createTime }
        for (fp in fpList) {
            val portionFood = foodMap[fp.foodId]
            require(portionFood != null) { "foodMap did not contain food with ID ${fp.foodId}" }
            fp.initFoodAndNd(portionFood)
            fp.servingId?.let {
                val serving = portionFood.getServingById(it)
                checkNotNull(serving) { "Serving specified by FoodPortion not found in its food!" }
                fp.initServing(serving)
            }
            fp.initMeal(meal)
            meal.addFoodPortion(fp)
        }
    }

    /* The getXXXBy(Id|Key) functions construct objects for all necessary entities that match the query,
     * as well as all other entities referenced by them.
     * For example, getMealsForDay constructs all of the MealTable objects for one particular day,
     * along with their FoodPortions, their Foods, and all of the Servings of those Foods.
     * It's probably worth caching the results of these!
     */
    @Throws(SQLException::class)
    fun getMealsById(ds: MacrosDatabase, mealIds: Collection<Long>): Map<Long, Meal> {
        if (mealIds.isEmpty()) {
            return emptyMap()
        }
        // Makes meal objects, filtering by the list of IDs. If mealIds is empty,
        // all meals will be returned.
        val meals = RawEntityQueries.getRawObjectsByKeys(ds, Meal.table, MealTable.ID, mealIds)
        val foodIds = getFoodIdsForMeals(ds, mealIds)
        if (foodIds.isNotEmpty()) {
            val foodMap = getFoodsById(ds, foodIds)
            for (meal in meals.values) {
                val foodPortionMap = getRawFoodPortionsForMeal(ds, meal)
                processRawFoodPortions(meal, foodPortionMap, foodMap)
            }
        }
        return meals
    }

    @Throws(SQLException::class)
    private fun getFoodIdsForMeals(ds: MacrosDatabase, mealIds: Collection<Long>): List<Long> {
        val ids = CoreQueries.selectSingleColumn(ds, FoodPortion.table, FoodPortionTable.FOOD_ID) {
            where(FoodPortionTable.MEAL_ID, mealIds)
            distinct()
        }
        // ensure no null IDs
        return ids.map { requireNotNull(it) { "Error: ID from database was null" }  }
    }

    @Throws(SQLException::class)
    fun getMealIdsForFoodIds(ds: MacrosDatabase, foodIds: Collection<Long>): List<Long> {
        return CoreQueries.selectNonNullColumn(ds, FoodPortion.table, FoodPortionTable.MEAL_ID) {
            where(FoodPortionTable.FOOD_ID, foodIds)
            distinct()
        }
    }

    @Throws(SQLException::class)
    fun getDaysForMealIds(ds: MacrosDatabase, mealIds: Collection<Long>): List<DateStamp> {
        return CoreQueries.selectNonNullColumn(ds, Meal.table, MealTable.DAY) {
            where(MealTable.ID, mealIds)
        }
    }
}